pageserver/src/layered_repository.rs

```diff
@@ -925,6 +930,69 @@ impl Timeline for LayeredTimeline {
             _write_guard: self.write_lock.lock().unwrap(),
         })
     }
+
+    fn find_record_lsn(
+        &self,
+        key: Key,
+        request_lsn: Lsn,
+        filter: &dyn Fn(ZenithWalRecord) -> bool,
+    ) -> Result<Option<(ZTimelineId, Lsn)>> {
```

Comment:
I have to agree with you.
Copying this layer processing logic is really ugly.
I also think how it is possible to refactor it (provide some kind of layer's iterator)
but I think it can be postponed for the future.